### ISSUE TITLE
Hysteresis upon "Restart Python interpreter"

### ISSUE THREAD

**Issue by spyder-issues-migrator** [spyder-ide/spyder#107]
_From [pybourgu...@gmail.com](https://code.google.com/u/113639270771538909286/) on 2009-11-07T11:18:37Z_

I am using ctypes ( http://docs.python.org/library/ctypes.html ) in spyder.
Restarting the python interpreter does not unload the libraries loaded
using ctypes. I am using Linux, Spyder version 1.0.1.

A simple procedure to reproduce it is the following:
1. Write a basic C file (single function that returns 1):
   int one() { return 1 ; }
2. Compile it with: gcc -shared -o one.so one.c
3. In spyder's python interpreter, type:
   from ctypes import *
   libone = cdll.LoadLibrary("path to one.so")
   libone.one()
This will return 1
4. Now modify the C file to return 2 instead of one, and recompile
5. Restart the python interpreter and reproduce step 3: 1 is still returned.
6. Relaunch spyder and reproduce step 3: 2 is returned.

_Original issue: http://code.google.com/p/spyderlib/issues/detail?id=107_

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2009-11-08T06:03:19Z_

Restarting Python interpreter in Spyder does not reload modules -- note that
reloading modules is not supported by the Python interpreter unless it is requested
explicitely with the 'reload' builtin function (and Interactive Console's 'restart
Python interpreter' in Spyder does not really restart the Python interpreter because
Spyder itself depends on it...).

The purpose of the 'Restart Python interpreter' feature is only to remove references
to alive objects (and free memory) -- not to unload completely modules (.pyc will be
loaded instead of .py -that's the way Python works- when re-importing modules after
restarting the interpreter).

In other words, what you should do to make it work is:
    * step 3:
import ctypes
libone = ctypes.cdll.LoadLibrary("....")
libone.one()
    * step 5:
instead of 'restart python interpreter':
del libone
reload(ctypes)

I'm just guessing, this might not work.

Anyway, as for issues related to reloading modules, please take a look at these: http://packages.python.org/spyder/console.html#reloading-modules https://code.google.com/p/spyderlib/issues/detail?id=62&can=1

**Status:** Done

**Comment by spyder-issues-migrator** [comment]
_From [pybourgu...@gmail.com](https://code.google.com/u/113639270771538909286/) on 2009-11-08T07:04:31Z_

Thanks for your prompt answer, which makes full sense. It is now clear to me that the
issue has nothing to do with Spyder itself.

The solution you suggest does not work, however, obviously because a C library loaded
by ctypes is neither a module nor a variable... The solution seems to rely on a
"FreeLibrary" instruction provided by ctypes, which is unfortunately not mentioned in
the documentation. I will dig this further, and report a complete solution to this
problem here as soon as I find it.

And by the way, Spyder is really a fantastic software! Thanks so much, I had been
looking for such an interface for years!
